src/client/client.rs

```diff
@@ -87,33 +86,38 @@ impl Connection {
     /// If the connection is not yet open, then a new connection will be opened.
     /// If the request cannot be written, then a new connection is opened and the request is retried once more.
     fn send(&mut self, request: &Request) -> Result<Response, RequestError> {
-        self.try_write(request)?;
-        read_response(self.reader.as_mut().unwrap()).map_err(ResponseParsingError::into)
+        self.try_write(request).map_err(|err| Sending(err))?;
+
+        let response_parser = ResponseParser::new();
+        match response_parser.parse(self.reader.as_mut().unwrap())? {
+            Done(response) => Ok(response),
+            IoErr(_, err) => Err(Reading(err))
+        }
     }
 
     /// Tries to write the request to the server.
     /// If an existing connection is open, then that connection will be written to, otherwise a new connection is opened.
     /// If the existing connection cannot be written to, then a new connection is opened.
-    fn try_write(&mut self, request: &Request) -> Result<(), RequestError> {
+    fn try_write(&mut self, request: &Request) -> Result<(), Error> {
         self.ensure_connected()?;
         if let Ok(_) = write_request(self.writer.as_mut().unwrap(), request) {
```

Comment:
Is this the best way to check if a fn returns ok? Genuinely curious since this syntax seems wordy

Reply:
Hm yeah it is a little wordy. I could use the .is_ok() method instead to make it more straightforward.

Reply:
That's what I was wondering about. There are so many functions like that I forget that they exist sometimes